packages/editor/src/components/local-autosave-monitor/index.js

```diff
@@ -0,0 +1,191 @@
+		const saveToSessionStorage = () => {
+			window.sessionStorage.setItem( postKey( postId ), JSON.stringify( {
+				post_title: getEditedPostAttribute( 'title' ),
+				content: getEditedPostAttribute( 'content' ),
+				excerpt: getEditedPostAttribute( 'excerpt' ),
+			} ) );
+		};
+		requestIdleCallback( saveToSessionStorage );
+	}, [ postId, getEditedPostAttribute ] );
+}
+
+/**
+ * Custom hook which manages the creation of a notice prompting the user to
+ * restore a local autosave, if one exists.
+ */
+function useAutosaveNotice() {
+	const {
+		postId,
+		getEditedPostAttribute,
+	} = useSelect( ( select ) => ( {
+		postId: select( 'core/editor' ).getCurrentPostId(),
+		getEditedPostAttribute: select( 'core/editor' ).getEditedPostAttribute,
+	} ) );
+
+	const { createWarningNotice, removeNotice } = useDispatch( 'core/notices' );
+	const { editPost, resetEditorBlocks } = useDispatch( 'core/editor' );
+
+	useEffect( () => {
+		let autosave = window.sessionStorage.getItem( postKey( postId ) );
+		if ( ! autosave ) {
+			return;
+		}
+
+		try {
+			autosave = JSON.parse( autosave );
+		} catch ( error ) {
+			// Not usable if it can't be parsed.
+			return;
+		}
+
+		const { post_title: title, content, excerpt } = autosave;
+		const edits = { title, content, excerpt };
+
+		// Only display a notice if there is a difference between what has been
+		// saved and that which is stored in sessionStorage.
+		const hasDifference = Object.keys( edits ).some( ( key ) => {
+			return edits[ key ] !== getEditedPostAttribute( key );
+		} );
+
+		if ( ! hasDifference ) {
+			// If there is no difference, it can be safely ejected from storage.
+			window.sessionStorage.removeItem( postKey( postId ) );
+
+			return;
+		}
+
+		const noticeId = uniqueId( 'wpEditorAutosaveRestore' );
+
+		createWarningNotice( __( 'The backup of this post in your browser is different from the version below.' ), {
+			id: noticeId,
+			actions: [
+				{
+					label: __( 'Restore the backup' ),
+					onClick() {
+						editPost( omit( edits, [ 'content' ] ) );
+						resetEditorBlocks( parse( edits.content ) );
+						removeNotice( noticeId );
+					},
+				},
+			],
+		} );
+	}, [ postId ] );
+}
+
+/**
+ * Custom hook which ejects a local autosave after a successful save occurs.
+ */
+function useAutosavePurge() {
+	const {
+		postId,
+		isDirty,
+	} = useSelect( ( select ) => ( {
+		postId: select( 'core/editor' ).getCurrentPostId(),
+		isDirty: select( 'core/editor' ).isEditedPostDirty(),
+	} ) );
+
+	const lastIsDirty = useRef( isDirty );
+
+	useEffect( () => {
+		if ( lastIsDirty.current && ! isDirty ) {
+			window.sessionStorage.removeItem( postKey( postId ) );
+		}
+
+		lastIsDirty.current = isDirty;
+	}, [ isDirty ] );
+}
+
+function LocalAutosaveMonitor() {
+	const autosave = useAutosaveCallback();
+	useAutosaveNotice();
+	useAutosavePurge();
+
+	return (
+		<AutosaveMonitor
+			interval={ AUTOSAVE_INTERVAL_SECONDS }
+			autosave={ autosave }
+			shouldThrottle
```

Comment:
Why do we pass `shouldThrottle` always set to `true`? Well, the question is if we really need all the logic which covers the case where it would be `false`.

Reply:
The point was to leave server-side autosave untouched. Presumably, since there's a network transaction and server-side execution, we'd want to be more conservative with autosave — hence defaulting to its debounced action. With local autosave we can be more eager. What do you think?

Reply:
I didn't notice we are reusing the existing component. It's all good. I wasn't sure why we even `shouldThrottle` have to pass this value in the first place. Now, it's all clear.